File: src/beans/interfaces/GraphBuilder.java
package beans.interfaces;

import beans.regles.Rule;
import guru.nidi.graphviz.attribute.Color;
import guru.nidi.graphviz.attribute.Rank;
import guru.nidi.graphviz.attribute.Rank.RankDir;
import guru.nidi.graphviz.attribute.Style;
import guru.nidi.graphviz.engine.Format;
import guru.nidi.graphviz.engine.Graphviz;
import guru.nidi.graphviz.model.Graph;
import guru.nidi.graphviz.model.Link;
import guru.nidi.graphviz.model.Node;
import org.jdom2.Element;

import java.io.File;
import java.io.IOException;
import java.util.*;
import java.util.stream.Collectors;

import static guru.nidi.graphviz.model.Factory.*;

public class GraphBuilder {
  private String name;
  private List<Node> nodes = new ArrayList<>();

  public GraphBuilder(String name) {
    this.name = name;
  }

  private void addNode(Node node) {
    this.nodes.add(node);
  }

  public void generate(Element rules, List<Rule> usedRules) {
    HashSet<String> ruleNames = usedRules.stream()
        .map(Rule::getName)
        .collect(Collectors.toCollection(HashSet::new));
    HashSet<String> trueFactNames = usedRules.stream()
        .map(rule1 -> rule1.getConclusion().name())
        .collect(Collectors.toCollection(HashSet::new));

    List<Element> children = rules.getChildren();
    Node firstNode = node(
        children.get(0)
            .getChild("premisses")
            .getChild("fait")
            .getAttributeValue("nom")
    ).with(Color.RED);
    addNode(firstNode);

    for (Element rule : rules.getChildren()) {
      String ruleName = rule.getAttributeValue("nom");
      Element premiseElements = rule.getChild("premisses");
      Element conclusion = rule.getChild("conclusion");
      String conclusionName = conclusion.getAttributeValue("nom");

      Node ruleNode = node(ruleName).with(Color.LIGHTBLUE);

      HashMap<Node, Link> linkToRuleByPremiseNode = new HashMap<>();
      for (Element premiseElement : premiseElements.getChildren()) {
        String premiseName = premiseElement.getAttributeValue("nom");
        Node premiseNode = node(premiseName);
        Link linkToRule = to(ruleNode);
        // Store ruleNode and a link toward it in a map in order to link them later
        // We cannot link them here because if the rule is verified the link must be in green
        // So, we modify the link in the part of the code which manage the case where the rule is verified
        // If we link here, the change on the link made here will not be taken in account (once the link is linked, it cannot be changed)
        linkToRuleByPremiseNode.put(premiseNode, linkToRule);
      }

      Node conclusionNode = node(conclusionName);
      Link linkToConclusion = to(conclusionNode);

      // Add colors
      if (ruleNames.contains(ruleName)) {
        // If the rule name is in the list of used rules
        ruleNode = ruleNode.with(Style.FILLED); // Fill used rules
        // Since the rule is used, its conclusion is true : color it in green
        conclusionNode = conclusionNode.with(Style.FILLED).with(Color.rgb(0, 255, 0));
        // Also color the link between the rule and its conclusion in green
        linkToConclusion = linkToConclusion.with(Color.rgb(0, 255, 0));
        // Also color links between premises and validated rule
        linkToRuleByPremiseNode.replaceAll((k, v) -> v.with(Color.rgb(0, 255, 0)));
      }

      for (Map.Entry<Node, Link> nodeLinkEntry : linkToRuleByPremiseNode.entrySet()) {
        Node premiseNode = nodeLinkEntry.getKey().link(nodeLinkEntry.getValue());
        addNode(premiseNode);
      }

      ruleNode = ruleNode.link(linkToConclusion);

      addNode(conclusionNode);
      addNode(ruleNode);
    }

  }

  public void export() throws IOException {
    Graph g = graph(name).directed()
        .graphAttr().with(Rank.dir(RankDir.TOP_TO_BOTTOM))
        .with(nodes);

    Graphviz.fromGraph(g).render(Format.DOT).toFile(new File(name + ".txt"));
    Graphviz.fromGraph(g).render(Format.PNG).toFile(new File(name + ".png"));
  }

}

File: src/beans/launches/GraphBuilder.kt
package beans.launches

import beans.regles.Rule
import org.jdom2.Element
import java.io.BufferedWriter
import java.io.File

fun BufferedWriter.writeln(string: String) {
    write("$string\n")
}

fun String.quote() = "\"$this\""


abstract class AbstractDotNode {
    protected val attrs = NodeAttrs()

    abstract override fun toString(): String

    fun put(key: String, value: String) {
        attrs[key] = value
    }
}


class DotNode(val name: String) : AbstractDotNode() {
    override fun toString(): String {
        return "${name.quote()} $attrs"
    }
}


class NodeAttrs : HashMap<String, String>() {
    override fun toString(): String {
        var string = "["
        for (entry in entries) {
            string += "${entry.key.quote()}=${entry.value.quote()}, "
        }
        string += "]"
        return string
    }
}

class DotLink(val first: DotNode, val second: DotNode) : AbstractDotNode() {
    override fun toString(): String {
        return "${first.name.quote()} -> ${second.name.quote()} $attrs"
    }
}


fun generateGraphAsDotCode(ruleElements: Element, verifiedRules: List<Rule>) {
    println("Build graph")
    val writer = File("graph.txt").bufferedWriter()
    writer.writeln("digraph {")

    val verifiedRulesNames = verifiedRules.map { it.name }
    val trueFacts = mutableSetOf<String>()
    for (verifiedRule in verifiedRules) {
        trueFacts.add(verifiedRule.conclusion.name())
        trueFacts.addAll(verifiedRule.premisses.map { it.name() })
    }

    val ruleNodes = mutableSetOf<DotNode>()
    val conclusionNodes = mutableSetOf<DotNode>()
    val premiseNodes = mutableSetOf<DotNode>()
    val links = mutableSetOf<DotLink>()

    for (ruleElement in ruleElements.children) {
        val ruleName = ruleElement.getAttributeValue("nom")
        val premiseElements = ruleElement.getChild("premisses")
        val conclusion = ruleElement.getChild("conclusion")
        val conclusionName = conclusion.getAttributeValue("nom")

        val ruleNode = DotNode(ruleName)
        ruleNodes.add(ruleNode)
        ruleNode.put("style", "filled")

        val conclusionNode = DotNode(conclusionName)
        conclusionNodes.add(conclusionNode)

        val ruleToConclusionLink = DotLink(ruleNode, conclusionNode)
        links.add(ruleToConclusionLink)

        for (premiseElement in premiseElements.children) {
            val premiseNode = DotNode(premiseElement.getAttributeValue("nom"))
            premiseNodes.add(premiseNode)
            val premiseToConclusionLink = DotLink(premiseNode, ruleNode)
            links.add(premiseToConclusionLink)
            premiseToConclusionLink.put("label", premiseElement.getAttributeValue("valeur") ?: "")
        }
    }

    for (ruleNode in ruleNodes) {
        if (ruleNode.name in verifiedRulesNames) {
            ruleNode.put("color", "red")
        } else {
            ruleNode.put("color", "lightblue")
        }
    }

    for (premiseNode in premiseNodes) {
        if (premiseNode.name in trueFacts) {
            premiseNode.put("color", "red")
        } else {
            premiseNode.put("color", "lightblue")
        }
    }

    for (conclusionNode in conclusionNodes) {
        if (conclusionNode.name in trueFacts) {
            conclusionNode.put("color", "red")
        } else {
            conclusionNode.put("color", "lightblue")
        }
    }

    for (link in links) {
        if (link.first.name in verifiedRulesNames || link.first.name in trueFacts) {
            link.put("color", "red")
        }
    }

    writer.writeln(ruleNodes.joinToString("\n"))
    writer.writeln(conclusionNodes.joinToString("\n"))
    writer.writeln(premiseNodes.joinToString("\n"))
    writer.writeln(links.joinToString("\n"))
    writer.write("}")
    writer.close()
}

fun generatePng(inputFile: File) {
    Runtime.getRuntime().exec("""dot -Tpng ${inputFile.absolutePath} -o outfile.png""").waitFor()
}
